src/dataloader/cache.rs

```diff
@@ -0,0 +1,209 @@
+use std::borrow::Cow;
+use std::collections::hash_map::RandomState;
+use std::collections::HashMap;
+use std::hash::{BuildHasher, Hash};
+use std::marker::PhantomData;
+use std::num::NonZeroUsize;
+
+/// Factory for creating cache storage.
+pub trait CacheFactory: Send + Sync + 'static {
+  /// Create a cache storage.
+  ///
+  /// TODO: When GAT is stable, this memory allocation can be optimized away.
```

Comment:
well, there is another point here to possible improvement - GAT is stable for a long time https://blog.rust-lang.org/2022/10/28/gats-stabilization.html

Reply:
I am not sure if I used this GAT thing properly but I did remove that `dyn` and made it statically type